### ISSUE TITLE
reverse lookups: return label for manually entered uri

### ISSUE THREAD

**Issue by pmgreen** [LD4P/sinopia_editor#1227]
Lookup label for manually entered URIs
Example: user pastes in 'http://id.loc.gov/authorities/names/n78093802' and gets 'Abbey, Edward, 1927-1989'

**Comment by michelleif** [comment]
intriguing. what would this be used for? maybe getting to larger question of where and when labels are stored.

**Comment by NancyJean** [comment]
@michelleif it would be used to display the label to humans (currently we only see the URI in preview if we manually entered a URI) and for searching on the label (which we cannot do now if we manually enter the URI) - so it would index the associated label for both display and search.

**Comment by michelleif** [comment]
@pmgreen @NancyJean : does this request stem from not being able to effectively find what you need in a QA lookup? in other words, in a perfect world if QA performed exactly how you expected, would you still need to enter a URI that exists in a lookup source? if this request is stemming from not being able to get the results you need when doing QA lookups, can you let us know (if you haven't already) which lookups are not working as expected in QA?
(including @elrayle and @sfolsom also, who would like to follow this discussion)

**Comment by NancyJean** [comment]
@michelleif  there are several use cases for this enhancement --
1) While we are waiting for a vocab to be fully usable/searchable in QA, we need to manually enter URIs. Currently, we need to manually add URIs because we cannot get MeSH URIs for subjects with valid $a$x combos via QA (MeSH RDF controls valid $a$x combos and issues URIs for them); however, I believe the QA team is already working on that or at least there already a ticket for it.  

2) I think this is a useful enhancement for vocabs that are not in QA at all.  I don't think it's realistic to assume that QA will be able to accommodate every standard vocab or authority that everyone may want to use.  This enhancement in Sinopia provide functionality for those (possibly) edge use cases or simply less mainstream vocabs or locally created vocabs.  For example, NLM might have an interest in SCORO, which is a very granular credit roles/contributor vocab that leans toward scientific research, if we want to use Sinopia for cataloging scientific datasets.  It's not likely there would be much use for that in traditional cataloging.

3) (related to 1) Even if a vocab is mainstream for a specific community/special library group, e.g., MeSH, it is not necessarily a development priority among the Cohort for QA.  The only reason MeSH works in QA at all right now is because despite not getting votes to prioritize development, Steven asked the Cohort if he could move it up the development list.

I'm not complaining -- just saying that we need to be pragmatic about expectations for QA and that this Sinopia enhancement allows for that.

**Comment by michelleif** [comment]
another example:
When doing a QA look-up, users have the option to manually enter a URL. QA should then be queried with URL to search for a label and if found persist the label to RDF. 
Pasting in http://id.nlm.nih.gov/mesh/D014780 into a subject look-up field, the URL will be persisted with an rdfs:label of Viruses.

**Comment by elrayle** [comment]
If I were going to do this through QA, I would do a fetch for the URI and get the label from the results.

Example:

URI:

https://lookup.ld4l.org/authorities/fetch/linked_data/locgenres_ld4l_cache?format=json&uri=http://id.loc.gov/authorities/genreForms/gf2011026047

Results:
```
{
  "uri": "http://id.loc.gov/authorities/genreForms/gf2011026047",
  "id": "gf2011026047",
  "label": [ "Animal films" ],
  "broader": [
    "http://id.loc.gov/authorities/genreForms/gf2011026406"
  ],
  ...
}
```

Extracted label: **`"Animal films"`**

**Comment by justinlittman** [comment]
Given a URI, how do we know what authority to use?

**Comment by sfolsom** [comment]
Is the idea that the cataloger would be working from a form in a field where the lookup has already been determined?

**Comment by michelleif** [comment]
in [NancyF's comment above ](https://github.com/LD4P/sinopia_editor/issues/1227#issuecomment-604533967) one of the use cases is for vocabs that are not in QA

**Comment by NancyL** [comment]
So there are really 2 different requirements here: 1. to enter a URI from an authority in QA and it returns the label. We do need to clarify Steven's question whether the lookup vocabulary is already determined in the field, or if there is a choice, or if it is not any vocabulary currently defined in that field in the template, and: 2. Enter a URI for an authority that is not in QA at all and add a label that is indexed (do I have that right?)

**Comment by michelleif** [comment]
#1 is: allow lookup to QA vocabs by URI. so if that is desired, let's make a ticket in QA repo.

**Comment by NancyJean** [comment]
Sorry, it took me a few minutes to find my account sign in info - ugh!

So, yes, I think that NancyL has it right.  In the first case, and perhaps limited interest, if whether QA is now able to get MeSH RDF main and 2ndary terms ($a$x).  In the 2nd case, it's if I give it a URI for a term that's not in a QA vocab, could QA use that URI to return the vocab term and keep the URI under the hood. 
Does that make sense?

**Comment by justinlittman** [comment]
Given an arbitrary URI, how would we get the label/readable term? URI's aren't necessarily resolvable and even if they are, there is no standard way of determining a label. Am I misunderstanding?

**Comment by elrayle** [comment]
For resolvable URIs, we can create a very simple QA config that allows fetch for a single URI only.  The config would define the LDPath for retrieving the label.

**Comment by michelleif** [comment]
NancyF's comment above: "Currently, we need to manually add URIs because we cannot get MeSH URIs for subjects with valid $a$x combos via QA (MeSH RDF controls valid $a$x combos and issues URIs for them); however, I believe the QA team is already working on that or at least there already a ticket for it."

Is there a QA issue for this?

**Comment by michelleif** [comment]
the use cases here will be solved by #2843, although the user will have to paste their own label in. closing this ticket.
